test/runner.rb - allow to load gems from install or build,
fix folder selection

---
 test/runner.rb | 19 +++++++++++++++----
 1 file changed, 15 insertions(+), 4 deletions(-)

diff --git a/test/runner.rb b/test/runner.rb
index 3e2e1316b1ab..585449fd83c1 100644
--- a/test/runner.rb
+++ b/test/runner.rb
@@ -1,11 +1,18 @@
 # frozen_string_literal: true
 
+require 'rbconfig'
+
 # Should be done in rubygems test files?
 ENV["GEM_SKIP"] = ENV["GEM_HOME"] = ENV["GEM_PATH"] = "".freeze
 
-# Get bundled gems on load path
-Dir.glob("#{__dir__}/../gems/*/*.gemspec")
-  .reject {|f| f =~ /minitest|test-unit|power_assert/ }
-  .map {|f| $LOAD_PATH.unshift File.join(File.dirname(f), "lib") }
+# Get bundled gems on load path, try install dir first
+gem_dir = "#{RbConfig::CONFIG['rubylibdir'].sub('/ruby/', '/ruby/gems/')}/gems"
+unless Dir.exist? gem_dir
+  gem_dir = "#{__dir__}/../gems"
+end
+
+Dir.glob("#{gem_dir}/*/lib")
+  .reject {|f| Dir.exist?(f) and f =~ /minitest|test-unit|power_assert/ }
+  .map {|f| $LOAD_PATH.unshift f }
 
 require_relative '../tool/test/runner'
